Compile cleanly / sort out sizeof(size_t) != sizeof(int)

ah- opened the pull request:
I'm trying to get librdkafka to compile 100% clean with msvc, clang and gcc.

This got longer than expected, so I would appreciate if you had a look.
It's not finished yet, but the remaining warnings are all about rather tricky issues.

Mostly replacing ints with size_t where the int was actually just assigned to the return value of something like strlen or where the value is passed to something like malloc.

Watch out for cases where I might have overlooked things that are now size_ts might actually be negative. I think I got them all now, but best to triple-check.

How would you prefer to continue with this? I think part of this PR should be fairly uncontroversial, another part needs checking by you, and then there's also a bunch of remaining warnings left.

Tests pass on win32/64 and with clang on osx, but haven't run them with valgrind yet.

edenhill commented:
+1 for effort!

I think it would make more sense to squash all these into one single commit, it is all the same thing and should be an atomic commit. It also makes reviewing simpler.

These changes needs to be verified across the board, including old CentOS:es, FreeBSD, Solaris 10, Solaris 11, various OSX versions, Cygwin.

edenhill commented:
Easiest way to squash the commits is probably to do:
`git rebase -i master` then mark all but the first one with `s`

edenhill commented on src/rdkafka_buf.c:
```diff
@@ -358,7 +358,7 @@ int rd_kafka_buf_write_Message (rd_kafka_buf_t *rkbuf,
 
 	/* Finalize Crc */
-	rd_kafka_buf_update_u32(rkbuf, of_Crc,
+	rd_kafka_buf_update_u32(rkbuf, (uint32_t) of_Crc,
```
The offset is size_t, why cast to uint32_t?

edenhill commented on src/rdkafka_buf.h:
```diff
@@ -39,11 +39,11 @@ typedef struct rd_kafka_broker_s rd_kafka_broker_t;
                         rd_kafka_buf_parse_fail(rkbuf,			\
 						"Not enough room in marshall buffer: " \
-						"%i+%i > %i",		\
+						"%zu+%zu > %zu",		\
```
From what I understand the only portable formattering for (s)size_t is `..%"PRIdsz"`

edenhill commented on src/rdkafka_cgrp.c:
```diff
@@ -1146,7 +1146,7 @@ static void rd_kafka_cgrp_terminate0 (rd_kafka_cgrp_t *rkcg, rd_kafka_op_t *rko)
         rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "CGRPTERM",
                      "Terminating group \"%.*s\" in state %s "
-                     "with %d partition(s)",
+                     "with %zu partition(s)",
```
%"PRIdsz"

edenhill commented on src/rdkafka_offset.c:
```diff
@@ -173,7 +173,7 @@ static int64_t rd_kafka_offset_file_read (rd_kafka_toppar_t *rktp) {
 	char *end;
 	int64_t offset;
-	int r;
+	size_t r;
```
the fread() below should probably be changed to:
if ((r = fread(buf, 1, ...)) < 1) {  rather than == -1

edenhill replied:
The `if ((r = fread(...)) == -1)` check can probably be removed completely. Just do  `r = fread(..);` and let the sub-sequent `r == 0` handle it (as is)

ah- replied:
Yeah, noticed that as well. It's unsigned anyways, so the -1 check doesn't make too much sense.

edenhill commented on src/rdkafka_range_assignor.c:
```diff
@@ -89,7 +89,7 @@ rd_kafka_range_assignor_assign_cb (rd_kafka_t *rk,
                  rd_kafka_dbg(rk, CGRP, "ASSIGN",
                               "range: Topic %s with %d partition(s) and "
-                              "%d subscribing member(s)",
+                              "%zu subscribing member(s)",
```
%"PRIdsz

edenhill commented on src/rdkafka_topic.c:
```diff
@@ -784,7 +784,7 @@ void rd_kafka_topic_partitions_remove (rd_kafka_itopic_t *rkt) {
         shptr_rd_kafka_toppar_t *s_rktp;
         shptr_rd_kafka_itopic_t *s_rkt;
-	int i;
+	size_t i;
```
Doesn't make sense to have this as explicit size_t, it just an iterator so keep it as int.

edenhill commented on src/rdlist.c:
```diff
@@ -31,27 +31,27 @@
 void rd_list_dump (const char *what, const rd_list_t *rl) {
-        int i;
-        printf("%s: (rd_list_t*)%p cnt %d, size %d, elems %p:\n",
+        size_t i;
```
same here, int

edenhill commented on src/rdlist.c:
```diff
@@ -31,27 +31,27 @@
-        int i;
-        printf("%s: (rd_list_t*)%p cnt %d, size %d, elems %p:\n",
+        size_t i;
+        printf("%s: (rd_list_t*)%p cnt %zu, size %zu, elems %zu:\n",
```
%"PRIdsz..

edenhill commented on src/rdlist.c:
```diff
@@ -82,7 +82,7 @@ static void rd_list_remove0 (rd_list_t *rl, int idx) {
 void *rd_list_remove (rd_list_t *rl, void *match_elem) {
         void *elem;
-        int i;
+        size_t i;
```
keep as int

edenhill commented on src/rdlist.c:
```diff
@@ -98,7 +98,7 @@ void *rd_list_remove (rd_list_t *rl, void *match_elem) {
                           int (*cmp) (void *_a, void *_b)) {
         void *elem;
-        int i;
+        size_t i;
```
keep as int

edenhill commented on src/rdlist.c:
```diff
@@ -127,7 +127,7 @@ void rd_list_destroy (rd_list_t *rl, void (*free_cb) (void *)) {
 
         if (rl->rl_elems) {
-                int i;
+                size_t i;
```
keep as int

edenhill commented on src/rdlist.c:
```diff
@@ -149,7 +149,7 @@ void *rd_list_elem (const rd_list_t *rl, int idx) {
 void *rd_list_find (const rd_list_t *rl, const void *match,
                     int (*cmp) (const void *, const void *)) {
-        int i;
+        size_t i;
```
keep as int

edenhill commented on src/rdlist.h:
```diff
@@ -36,8 +36,8 @@
-        int    rl_cnt;
+        size_t rl_size;
+        size_t rl_cnt;
         void **rl_elems;
```
Not sure this is warranted?

edenhill commented on src/rdlist.h:
```diff
@@ -125,7 +125,7 @@ void *rd_list_elem (const rd_list_t *rl, int idx);
  * Returns the number of elements in list.
  */
-static __inline RD_UNUSED int rd_list_cnt (const rd_list_t *rl) {
+static __inline RD_UNUSED size_t rd_list_cnt (const rd_list_t *rl) {
```
return int, anything else is just messy imho

ah- commented:
Cool, thanks. This is exactly the early feedback I was looking for, will fix and rebase once I'm done.

ah- commented:
Updated, rebased and squashed.

There's two warnings left:

``` rdrand.c(40): warning C4267: '=' : conversion from 'size_t' to 'int', possible loss of data```
is due to rd_jitter only returning 32bit ints, while it tries to shuffle a potentially 64bit long list.

```rdkafka_broker.c(1082): warning C4244: '=' : conversion from 'SOCKET' to 'int', possible loss of data```
not quite sure how to fix that cleanly.

ah- commented on src/rdcrc32.h:
```diff
@@ -70,7 +70,7 @@ rd_crc32_t rd_crc32_reflect(rd_crc32_t data, size_t data_len);
 static __inline rd_crc32_t rd_crc32_init(void)
 {
-#if WITH_ZLIB
+#if WITH_ZLIB && !defined(_MSC_VER)
```
Due to zlib crc32 only running on uInt, which is 32 bit.

edenhill replied:
I dont really understand what you are trying to achieve here.
Are you trying to disable the use of zlib on win32? If so you should remove WITH_ZLIB from src/win32_config.h, but that seems questionable.

ah- replied:
Make it work with all possible inputs. It currently silently gives you the wrong crc32 if you give it a buffer that's >2 GB.

But I suppose that's not going to happen, I'll change it to an assert < 2GB and use zlib.

ah- commented on win32/librdkafka.vcxproj:
```diff
@@ -92,7 +92,7 @@
       <PrecompiledHeader>NotUsing</PrecompiledHeader>
       <WarningLevel>Level3</WarningLevel>
-      <PreprocessorDefinitions>WIN32;_DEBUG;_WINDOWS;_USRDLL;LIBRDKAFKA_EXPORTS;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions>WIN32;_DEBUG;_WINDOWS;_USRDLL;LIBRDKAFKA_EXPORTS;_CRT_SECURE_NO_WARNINGS;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
To disable warnings about strncpy being unsafe. It's used properly and strncpy_s isn't available on all platforms.

ah- commented on win32/librdkafka.vcxproj:
```diff
@@ -225,7 +225,9 @@
     <ClCompile Include="..\src\rdrand.c" />
-    <ClCompile Include="..\src\snappy.c" />
+    <ClCompile Include="..\src\snappy.c">
+      <ExcludedFromBuild>true</ExcludedFromBuild>
```
Since it's unused on win32 anyways and causes a couple of warnings.

edenhill commented:
re rdrand.c:40 - just cast as necessary.

re SOCKET to int - I'll fix that after the merge.

ah- commented:
Rebased onto master and updated to just cast for rdrand.c and crc32.

edenhill commented:
Big thanks, @ah-, great work!
